Create activity button component frontend unit tests

aishwary023 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes or fixes part of #NA.
2. This PR does the following: 
- Adds frontend unit tests for `create-activity-button.component.ts`

## Essential Checklist

- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The linter/Karma presubmit checks have passed locally on your machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
  - This lets reviewers restart your CircleCI tests for you.
- [x] The PR is made from a branch that's **not** called "develop".

## Proof that changes are correct

![image](https://user-images.githubusercontent.com/26626415/119961277-7ce76700-bfc3-11eb-9624-e4f2f6d79f8a.png)

## PR Pointers

- Make sure to follow the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Contributing-code-to-Oppia#instructions-for-making-a-code-change).
- Oppiabot will notify you when you don't add a PR_CHANGELOG label. If you are unable to do so, please @-mention a code owner (who will be in the Reviewers list), or ask on [Gitter](https://gitter.im/oppia/oppia-chat).
- For what code owners will expect, see the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia%27s-code-owners-and-checks-to-be-carried-out-by-developers).
- Make sure your PR follows conventions in the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide), otherwise this will lead to review delays.
- Never force push. If you do, your PR will be closed.
- Oppiabot can assign anyone for review/help if you leave a comment like the following: "{{Question/comment}} @{{reviewer_username}} PTAL"
- Some of the e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-your-build-fails).

aishwary023 committed:
Added unit tests for create-activity-button component and refactored init function

aishwary023 committed:
Fixed lint errors

aishwary023 committed:
Removed file name from NOT_FULLY_COVERED_FILENAMES

kevintab95 commented on core/templates/components/button-directives/create-activity-button.component.ts:
```diff
@@ -82,14 +82,15 @@ export class CreateActivityButtonComponent implements OnInit {
    }
-   ngOnInit(): void {
+
+   async ngOnInit(): Promise<void> {
```
This function does not seem like it returns anything, right?

kevintab95 replied:
Also, more generally, avoid making changes to source files for frontend test coverage PRs unless it is absolutely necessary.

aishwary023 replied:
Yep, it does not.

I added the changes as, `canCreateCollections` was not being set to `true` before the control reached [this](https://github.com/oppia/oppia/blob/32627e423d32b015720e2c1009fd4eca0aad04dc/core/templates/components/button-directives/create-activity-button.component.ts#L97) line hence inside the ngOnInit function, the value of `canCreateCollections` was false and we were not able to enter the else block.

Is there any other way to test this code, without using `await`?

kevintab95 replied:
However, I think you might have uncovered a possible bug in the code.

getUserInfoAsync() only sets `canCreateCollections` after the backend call completes. But we seem to use `canCreateCollections` outside the success callback so it could execute even if the call does not complete.

To answer your question, yep, it is possible. You'll just need to move more of the logic in to the success callback of the async function.

```ts
    this.userService.getUserInfoAsync().then(userInfo => {
      this.canCreateCollections = userInfo.canCreateCollections();
      this.userIsLoggedIn = userInfo.isLoggedIn();
      // If the user clicked on a 'create' button to get to the dashboard,
      // open the create modal immediately (or redirect to the exploration
      // editor if the create modal does not need to be shown).
      if (this.urlService.getUrlParams().mode === 'create') {
        if (!this.canCreateCollections) {
          this.explorationCreationService.createNewExploration();
        } else {
          this.initCreationProcess();
        }
      }
     });
```

That said, I'm fine with the current approach you have. It does the same thing and looks better. :)

srijanreddy98 replied:
async ngOnInit is not a good practice (this is perfectly fine code, but its counter intuitive). Please undo this change. Angular calls this function before the view renders,  and it won't wait for the promise to resolve. Using async await will make error handling a pain here. Also, this will cause the life cycle hooks to end in different order.

Here is some more context: https://github.com/angular/angular/issues/17420

kevintab95 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,279 @@
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(component.initCreationProcess).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(true);
+    expect(component.userIsLoggedIn).toBe(true);
```
Do we want a pre-check for these before `ngOnInit` is called?

aishwary023 replied:
Yes, that would be better. Added pre-checks.

kevintab95 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,279 @@
+    expect(userServiceSpy).toHaveBeenCalled();
+    expect(urlParamsSpy).toHaveBeenCalled();
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+    expect(component.canCreateCollections).toBe(false);
+    expect(component.userIsLoggedIn).toBe(true);
```
Ditto: Please add pre-checks.

aishwary023 replied:
Done

kevintab95 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,279 @@
+  it('should return from initCreateProcess() when' +
+    ' creation is in progress', () => {
+    component.creationInProgress = true;
+    expect(component.initCreationProcess()).toBe();
```
What are we expecting the output to be here? Also, do we need a pre-check?

aishwary023 replied:
We are expecting it to return void. I don't think we need a pre-check here. We are just checking if the function terminates when `creationInProgress` is `true` [here](https://github.com/oppia/oppia/pull/12936/files#diff-18b57c9d65ff111b9d10b87671ac5aadd027ffeaf361d8e341ad4342c7f05f98L61-L63)

kevintab95 replied:
Ah I see, thanks!

seanlip replied:
It's not clear to me that the return value being undefined corresponds to the creation not happening. Note that it should not be necessary to refer to the actual code in order to understand the test.

Can you write this a different way? You might need to tweak some stuff in the main file to signal whether initCreationProcess() succeeded or failed. Or you might have a test that checks the side effects (e.g. if initCreationProcess() calls some creation method you can check that it is not called). In either case you should also have a "pre-check" in this test that verifies that the behaviour in question works differently when there are no other creations in progress.

aishwary023 replied:
Changed the test.

kevintab95 reviewed (changes requested):
Thanks @aishwary023! Left some comments. PTAL

aishwary023 committed:
Addressed review comments

kevintab95 reviewed (approved):
Thanks @aishwary023! LGTM

aks681 reviewed (approved):
Lgtm!

DubeySandeep reviewed (approved):
LGTM for the codeowner file `scripts/check_frontend_test_coverage.py`

aishwary023 committed:
Addresed review comment

aishwary023 committed:
Fix Lint error

aishwary023 commented:
@srijanreddy98 - Thanks for the explanation! I removed `async ngOnInit` and moved the logic inside the success callback as mentioned by Kevin [here](https://github.com/oppia/oppia/pull/12936#discussion_r641859602).

nithusha21 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,285 @@
+      // suppress this error because we need only "isLoggedIn" and
+      // "canCreateCollections" functions for testing.
+      // @ts-expect-error
+      .and.returnValue(Promise.resolve({
```
Can we resolve to a full UserInfo object? that might be better

aishwary023 replied:
Added 2 types of user info - one who can create collections and one who cannot.

nithusha21 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,285 @@
+      _hash: '',
+      _hashChange: null,
+      _href: '',
+      get hash() {
```
Do you need all of these functions in the mock?

aishwary023 replied:
Nope. Removed unused functions.

nithusha21 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,285 @@
+    expect(component).toBeDefined();
+  });
+
+  it('should initialize and call initCreationProcess()', fakeAsync(() => {
```
Need to work on the names... Don't try to name it based on the code... Try to think about it more like a user. What this checks is basically that all the initializations happen.

Maybe rename this to:
should initialize component properties and then begin creation process

(Try to work on the names for the other suites, and I'll take a look again)

aishwary023 replied:
Renamed the tests, PTAL!

nithusha21 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,285 @@
+    expect(component.initCreationProcess()).toBe();
+  });
+
+  it('should create new exploration if user cannot create collections', () => {
```
This is a good test name

nithusha21 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,285 @@
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should replace location', () => {
```
Names should be of the form 

should do something if/when some "trigger" happens

aishwary023 replied:
Renamed

nithusha21 commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,285 @@
+    const urlServiceSpy = spyOn(urlService, 'getPathname').and.returnValue(
+      '/creator-dashboard');
+    const modalSpy = spyOn(ngbModal, 'open').and.callFake((dlg, opt) => {
+      setTimeout(opt.beforeDismiss);
```
Why do you need this? In the past timeouts in the code have resulted in flaky tests... Just want to make sure this is needed, before adding it.

aishwary023 replied:
Yep, we don't need that, moreover, `opt.beforeDismiss` is undefined in this case. Removed setTimeout. 

(We are using this at many other places in the codebase, may need to check them too once)

aishwary023 committed:
Addressed review comments

aishwary023 committed:
Fix lint err

aishwary023 commented:
@srijanreddy98 @nithusha21 Addressed the review, PTAL!
@seanlip PTAL! 
Thanks!

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+  }
+}
+
+interface UrlParamsType {
```
Is this a general construct? If so, why isn't it defined somewhere more central? /cc @srijanreddy98

srijanreddy98 replied:
Two comments here:
1> This is present in UrlService. Why not just export it from there and import here.
2> Can't this just be `Record<string, string>`? Why do we need an interface for this? https://www.typescriptlang.org/docs/handbook/utility-types.html#recordkeystype

aishwary023 replied:
I have used the 1st option

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+  let windowRef: MockWindowRef;
+  let ngbModal: NgbModal;
+
+  let userInfoCanCreateCollection = {
```
This is a bit awkwardly worded -- perhaps userInfoForCollectionCreator, userInfoForNonCollectionCreator perhaps.

aishwary023 replied:
Done

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Should create what ...?

Please have the describe/it blocks together be complete sentences of the form "X should...", where X is the thing being described.

aishwary023 replied:
Done

(I added the test name as it was in the [angular testing guide](https://angular.io/guide/testing-components-basics#beforeeach))

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    expect(component).toBeDefined();
+  });
+
+  it('should initialize and begin creation process if user can' +
```
What is the difference between "initialize" and "begin"?

What is the "creation process" that's being referred to here? Could you explain it in simpler terms? From what I can see, the test is checking that various random things have been called, and I don't see how that relates to the description of the test. Try to have a more specific description that matches what is being checked for in the implementation of the test.

aishwary023 replied:
Reworded the test description

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
```
What is the point of this check and the one on the next line (from a "testing the interface" perspective)?

aishwary023 replied:
Thanks for the comment. I was trying to test every call that is being made, to be sure that everything works. But, I got the point that we should only test the interface. Removed the lines.

seanlip replied:
Yup, note that testing calls is a bit of an antipattern. It often results in tests that are too interface-specific. See a similar comment I left on @Radesh-kumar's PR: https://github.com/oppia/oppia/pull/12944#discussion_r643709534

aishwary023 replied:
Noted, thanks!

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should initialize and create new exploration if user cannot' +
```
Perhaps: it should create a new exploration automatically if the user cannot create collections

aishwary023 replied:
Done

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    tick();
+    fixture.detectChanges();
+
+    expect(userServiceSpy).toHaveBeenCalled();
```
Why do we need this and the next line?

aishwary023 replied:
Removed

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    expect(component.initCreationProcess()).toBe();
+  });
+
+  it('should create new exploration if user cannot create collections', () => {
```
How is this different from https://github.com/oppia/oppia/pull/12936/files#diff-c6bbc8b315fa5210ba8944e75fc8abb35cca1ca8b84b8e074ab017d687aeffccR173 ?

aishwary023 replied:
There is a problem with the code. When `canCreateCollections` is `true` [here](https://github.com/oppia/oppia/blob/f753ff5929ca70a32b2fc22d5f9edb0dec8d8ccc/core/templates/components/button-directives/create-activity-button.component.ts#L97), then `initCreationProcess()` is called, but [this](https://github.com/oppia/oppia/blob/f753ff5929ca70a32b2fc22d5f9edb0dec8d8ccc/core/templates/components/button-directives/create-activity-button.component.ts#L68) line is executed only if `canCreateCollections` is false.

So, I had to check `initCreationProcess` separately, with `canCreateCollections = false`. Also, changed the name of the other tests to denote that it begins `activity` ceation.

seanlip replied:
Why not fix the code? That's kind of the point of doing the tests -- if the tests uncover a bug, then fix the bug, rather than leave the bug there and try to test a pathway that can't occur in practice.

aishwary023 replied:
I may have misjudged the code, this pathway is possible, when `initCreationProcess` is not called from `ngOnInt` but by clicking on the create button.

Regarding the comment above, I tried to completely remove the functionality to create collections, but then it was going to affect a lot of e2e tests as well. So, after discussion with @kevintab95 we decided to do that in a separate PR.

seanlip replied:
OK, but then I go back to the earlier question ... how is this different from the already-existing similar test above?

aishwary023 replied:
- This test checks the condition when - user is on the creator dashboard page & cannot create collections.
- The test above checks the condition when - user is redirected to creator dashboard after they log in & cannot create collections.

Modified the test description to state the same.

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    expect(modalSpy).toHaveBeenCalled();
+  });
+
+  it('should close modal when user clicks outside of' +
```
This is a good test statement but the test doesn't check this. You need to have some kind of check that the modal is actually closed (and a similar check before that verifies that the modal is open before the action is taken).

aishwary023 replied:
I think, only the success and rejection callback of the modal is needed to be tested (which is happening), but the test description is different. Should I change the test description accordingly or is testing whether a modal is opened or closed the right way to go?

seanlip replied:
As long as the test description and implementation match, and it tests something useful (that could conceivably go wrong and relates to what an external observer sees rather than an implementation detail) then things should be fine.

aishwary023 replied:
Changed the description.

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,290 @@
+    expect(replaceSpy).toHaveBeenCalledWith('/creator-dashboard?mode=create');
+  });
+
+  it('should open a create activity modal if user' +
```
You should have parity in the checks for similar tests. E.g. for this one there should be a check that replaceSpy is not called.

aishwary023 replied:
Done

seanlip replied:
Sorry -- was this description changed from before? Now this test isn't testing the opposite of the previous test (based on the description) so I would withdraw the replaceSpy comment. Though, in that spirit, it might make sense to have a separate test along the lines of: "it should not redirect user when user clicks create button and is on creator dashboard page".

aishwary023 replied:
Those might essentially be the same tests, only differing by the replaceSpy line. I have changed the description to `should not redirect user but open a create activity modal if user can create collections and is on creator dashboard page` to accommodate both tests.

seanlip reviewed (changes requested):
Thanks @aishwary023, took a pass.

Also, if it makes things simpler -- I think it is fine to remove the functionality for creating collections, so that clicking the Create button always automatically creates a new exploration. Up to you whether you want to do this in the PR -- I'm mentioning it since it may make the test logic simpler.

aishwary023 committed:
Modified tests to address review comments

aishwary023 committed:
Addressed review comment

aishwary023 committed:
Lint fix

aishwary023 committed:
Fix karma config

aishwary023 committed:
Fix frontend test

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,291 @@
+    tick();
+    fixture.detectChanges();
+
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
```
Nit: Conceptually, it may make more sense for these three expectations to be in reverse order: First, the user is logged in. Then, we find that they can't create collections. Finally, we confirm that the exploration creation service has been called.

I.e. try to have  each test tell a coherent story. That makes the test easier to follow.

aishwary023 replied:
Yep, that sounds better. Changed

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,291 @@
+    expect(component.userIsLoggedIn).toBe(true);
+  }));
+
+  it('should stop the creation process if another' +
```
More like "should not start a new creation process" right? (The current phrasing suggests stopping the existing one.)

aishwary023 replied:
Done

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,291 @@
+    component.initCreationProcess();
+    expect(explorationCreationServiceSpy).toHaveBeenCalledTimes(1);
+
+    component.creationInProgress = true;
```
Just to check, shouldn't this be an expectation (wouldn't it be expected to happen as a consequence of calling initCreationProcess())?

aishwary023 replied:
Yep, changed it to an expectation

aishwary023 committed:
Addressed reviews

aishwary023 committed:
Merge remote-tracking branch 'upstream/develop' into create-activity-button-tests

aishwary023 commented:
@seanlip -- addressed the comments. PTAL, thanks

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,289 @@
+    component.initCreationProcess();
+
+    expect(replaceSpy).not.toHaveBeenCalled();
+    expect(modalSpy).toHaveBeenCalled();
```
Should there be a parallel check for this in the previous test?

aishwary023 replied:
Yes, added

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,289 @@
+  }));
+
+  it('should show upload exploration modal', () => {
+    const explorationCreationServiceSpy = spyOn(
```
Name of spy does not match name of function.

aishwary023 replied:
Changed

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -0,0 +1,289 @@
+    expect(explorationCreationServiceSpy).toHaveBeenCalled();
+  });
+
+  it('should redirect to login page when user creates exploration' +
```
You'd probably want to add another test after this to verify that the method on siteAnalyticsService is called. That's an example of an external side effect that should be validated.

aishwary023 replied:
Done

seanlip replied:
No, I explicitly asked for a separate test here because you are testing a separate thing. This test checks redirection. A separate test should check analytics.

Again, have the test body match the test description. This is a very important fundamental point.

aishwary023 replied:
Ah, alright. Added another test

seanlip reviewed (changes requested):
Looks great @aishwary023, this is very close to being done. Thanks!

seanlip commented:
@srijanreddy98 @nithusha21 PTAL. Thanks!

aishwary023 committed:
Merge remote-tracking branch 'upstream/develop' into create-activity-button-tests

aishwary023 committed:
Addressed review

aishwary023 committed:
Addressed review

aishwary023 committed:
lint fix

srijanreddy98 reviewed (approved):
LGTM for url.service.ts.

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -146,8 +150,8 @@ describe('CreateActivityButtonComponent', () => {
-  it('should begin activity creation process if user can' +
-    ' create collections', fakeAsync(() => {
+  it('should begin activity creation process if user can create collections' +
+    ' and is redirected to creator dashboard', fakeAsync(() => {
```
Here's the issue I have with this, though. If you look at the body of this test, there is no mention of redirection to creator dashboard. So I think your test description is "breaking out of the boundaries" here. It should be self-contained with regards to the body of the test (i.e. the test body should obviously be setting up the conditions needed and then executing the test, and it should be easy to tie the test body back to the description).

Try again?

aishwary023 replied:
Made some changes, PTAL

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -165,8 +169,9 @@ describe('CreateActivityButtonComponent', () => {
 
-  it('should create a new exploration automatically if the user' +
-    ' cannot create collections', fakeAsync(() => {
+  it('should create a new exploration automatically if the user cannot create' +
```
This one, too -- I don't see anything in the test itself about redirection, either asserting it or setting it up etc., so it's hard for me to understand how the body of the test actually tests the thing in the description (which is the main thing the reviewer is looking for).

aishwary023 replied:
Made changes

aishwary023 committed:
Restructed tests, addressed review comment

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -150,44 +150,127 @@ describe('CreateActivityButtonComponent', () => {
+    }));
+
+    it('should create new exploration if the user is on' +
+      ' the creator dashboard page', () => {
```
Which part of this implementation guarantees that "the user is on the creator dashboard page"?

aishwary023 replied:
Changed the test description

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -150,44 +150,127 @@ describe('CreateActivityButtonComponent', () => {
 
-    expect(component.userIsLoggedIn).toBe(false);
-    expect(component.canCreateCollections).toBe(false);
+    it('should handle modal\'s failure callback' +
```
it should stop the creation process when there is an error with the XXX modal opening

(or similar -- "handle modal's failure callback is a bit too 'in the weeds'", and "handle" is not a specific word and should generally be avoided)

aishwary023 replied:
Changed

seanlip commented on core/templates/components/button-directives/create-activity-button.component.spec.ts:
```diff
@@ -150,44 +150,127 @@ describe('CreateActivityButtonComponent', () => {
+      component.canCreateCollections = true;
+      spyOn(urlService, 'getPathname').and.returnValue(
+        '/creator-dashboard');
+      const modalSpy = spyOn(ngbModal, 'open').and.callFake((dlg, opt) => {
```
Can this spy on a specific modal (or be applied to a specific modal) rather than all modals in general?

aishwary023 replied:
modalSpy can only spy on all modal in general, but we can check if it was called with a specific spy. Changed the test

seanlip reviewed (changes requested):
Thanks -- the refactoring is better. Left a few more comments.

seanlip commented:
@nithusha21 PTAL as codeowner

aishwary023 committed:
Addressed review comment

seanlip reviewed (approved):
LGTM. Thanks @aishwary023 (and well done)!

nithusha21 reviewed (approved):
Looks great! Thanks Aishwary
